quickwit/quickwit-proto/src/lib.rs

```diff
@@ -287,10 +289,105 @@ impl<'a> Extractor for MetadataMap<'a> {
     }
 }
 
-
 /// Sets parent span context derived from [`tonic::metadata::MetadataMap`].
 pub fn set_parent_span_from_request_metadata(request_metadata: &tonic::metadata::MetadataMap) {
     let parent_cx =
-    global::get_text_map_propagator(|prop| prop.extract(&MetadataMap(request_metadata)));
+        global::get_text_map_propagator(|prop| prop.extract(&MetadataMap(request_metadata)));
     Span::current().set_parent(parent_cx);
 }
+
+impl ToString for IndexingTask {
+    fn to_string(&self) -> String {
+        format!(
+            "{}:{}:{}",
+            self.index_id, self.source_id, self.incarnation_id
+        )
+    }
+}
+
+impl TryFrom<&str> for IndexingTask {
+    type Error = anyhow::Error;
+
+    fn try_from(index_task_str: &str) -> anyhow::Result<IndexingTask> {
+        let mut iter = index_task_str.split(':');
+        let index_id = iter.next().ok_or_else(|| {
+            anyhow!(
+                "Invalid index task format, cannot find index_id in `{}`",
+                index_task_str
+            )
+        })?;
+        let source_id = iter.next().ok_or_else(|| {
+            anyhow!(
+                "Invalid index task format, cannot find source_id in `{}`",
+                index_task_str
+            )
+        })?;
+        let incarnation_id = iter.next().map_or_else(
```

Comment:
nitpick:
Rust std has a massive number of functional methods. I personally think 
the most obscure ones just make the code less readable. I'd put `.map_or_else(...)` into that category. 


Here `map` and `unwrap_or_else` would work.

My personal favorite, especial when Error is involved like this, is to just use an if statement.

```rust
let incarnation_id = 
     if let Some(incarnation_id_str) = iter.next() {
           Ulid::from_string(encoded)
                 .context("Invalid index task format, cannot parse incarnation_id")?
     } else {
           Ulid::default()
     };
 ```